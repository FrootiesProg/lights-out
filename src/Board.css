.Board {
  display: flex;
  flex-wrap: wrap;
  max-width: 380px;
  margin: 0 auto;
  background-color: #333;
  border-radius: 8px;
  padding: 10px;
}

.board-row {
  display: flex;
}
