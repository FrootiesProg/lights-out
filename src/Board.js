import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard());

  function createBoard() {
    const initialBoard = Array.from({ length: nrows }, () =>
      Array.from({ length: ncols }, () => Math.random() < chanceLightStartsOn)
    );
    return initialBoard;
  }

  function flipCell(y, x, board) {
    if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
      board[y][x] = !board[y][x];
    }
  }

  function flipCellsAround(coord) {
    const [y, x] = coord.split("-").map(Number);

    const boardCopy = board.map((row) => [...row]);

    flipCell(y, x, boardCopy);
    const adjacentCells = [
      [y, x - 1],
      [y, x + 1],
      [y - 1, x],
      [y + 1, x],
    ];
    for (const [adjY, adjX] of adjacentCells) {
      flipCell(adjY, adjX, boardCopy);
    }

    setBoard(boardCopy);
  }

  function hasWon() {
    return board.every((row) => row.every((cell) => !cell));
  }

  if (hasWon()) {
    return <div className="winner">You won!</div>;
  }

  let tableBoard = [];
  for (let y = 0; y < nrows; y++) {
    let row = [];
    for (let x = 0; x < ncols; x++) {
      let coord = `${y}-${x}`;
      row.push(
        <Cell
          key={coord}
          isLit={board[y][x]}
          flipCellsAroundMe={() => flipCellsAround(coord)}
        />
      );
    }
    tableBoard.push(
      <tr key={y} className="board-row">
        {row}
      </tr>
    );
  }

  return <div className="Board">{tableBoard}</div>;
}

export default Board;
