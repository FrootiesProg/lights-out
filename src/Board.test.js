import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Board from "./Board";

test("renders the starter board", () => {
  const { asFragment } = render(
    <Board nrows={5} ncols={5} chanceLightStartsOn={0.25} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("handles cell-clicking", () => {
  render(<Board nrows={5} ncols={5} chanceLightStartsOn={0.25} />);

  const firstCell = screen.getByTestId("cell-0-0");
  userEvent.click(firstCell);

  // Verify that the clicked cell and its neighbors have flipped
  expect(firstCell).toHaveClass("Cell-lit");
  //  Check other cells as needed
});

test("displays 'You won!' message on win", () => {
  const winBoard = [
    [false, false, false],
    [false, false, false],
    [false, false, false],
  ];

  render(<Board nrows={3} ncols={3} chanceLightStartsOn={0.25} />);
  // Override the board state with a winning configuration
  jest.spyOn(Board.prototype, "createBoard").mockReturnValue(winBoard);

  // Verify that the "You won!" message is displayed
  expect(screen.getByText("You won!")).toBeInTheDocument();
});
