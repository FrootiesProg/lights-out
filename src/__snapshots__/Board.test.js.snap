// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the starter board 1`] = `
<DocumentFragment>
  <div
    class="Board"
  >
    <div
      class="board-row"
    >
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell Cell-lit"
      />
      <div
        class="Cell "
      />
      <div
        class="Cell Cell-lit"
      />
    </div>
    <div
      class="board-row"
    >
      <div
        class="Cell "
      />
      <div
        class="Cell Cell-lit"
      />
      <div
        class="Cell Cell-lit"
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
    </div>
    <div
      class="board-row"
    >
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell Cell-lit"
      />
    </div>
    <div
      class="board-row"
    >
      <div
        class="Cell "
      />
      <div
        class="Cell Cell-lit"
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
    </div>
    <div
      class="board-row"
    >
      <div
        class="Cell Cell-lit"
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
      <div
        class="Cell "
      />
    </div>
  </div>
</DocumentFragment>
`;
